<launch>
    <!-- k3dom_demo -->
    <!-- Sets up an k3dom node with parameters specified by args -->
    <!-- See yaml files at k3dom/config/methods/ for config parameters -->
    <!-- method: ds_k3dom or k3dom or dsphdmib -->
    <!-- dataset: simple or complex -->
    <arg name="method" default="ds_k3dom"/>
    <arg name="dataset" default="experiment"/>

    <arg name="lidar_topic" default="/velodyne_points"/>
    <arg name="lidar_frame" default="velodyne"/>
    <arg name="pose_topic" default="/pose"/>

    <arg name="config_dir" value="$(find ds_k3dom)/config"/>
    <param name="/use_sim_time" value="true"/>
    
    <!--Replay data-->
    <arg name="bag_dir" default="$(find ds_k3dom)/data"/>
    <arg name="rate_factor" default="0.3"/>
    <arg name="delay" default="3"/>
    <arg name="start" default="0"/>
    <node pkg="rosbag" type="play" name="player1" output="screen" args="-d $(arg delay) -s $(arg start) -r $(arg rate_factor) --clock $(arg bag_dir)/$(arg dataset).bag"/>

    <!--Start the k3dom node -->
    <!-- for debug, launch-prefix="xterm -e gdb ~~args"  !note change ~~ to double hypen -->
    <node pkg="ds_k3dom" type="$(arg method)_node" name="$(arg method)_node" output="screen">

        <param name="lidar_topic" value="$(arg lidar_topic)"/>
        <param name="lidar_frame" value="$(arg lidar_frame)"/>
        <param name="pose_topic" value="$(arg pose_topic)"/>

        <!-- Load method configuration from file -->
        <rosparam file="$(arg config_dir)/$(arg method).yaml" command="load"/>

        <!-- Load dataset configuration from file -->
        <rosparam file="$(arg config_dir)/$(arg dataset).yaml" command="load"/>
        
    </node>
    
    <!-- tf -->
    <node pkg="tf2_ros" type="static_transform_publisher" name="lidar" args="0 0 0 0 0 0 base_link $(arg lidar_frame)"/>
    <node pkg="tf2_ros" type="static_transform_publisher" name="map" args="0 0 0 0 0 0 global_map map"/>

    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find ds_k3dom)/rviz/$(arg dataset).rviz"/>
</launch>
